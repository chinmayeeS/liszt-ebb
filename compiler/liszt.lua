-- Imports
package.path = package.path .. ";./compiler/?.lua;./compiler/?.t"
require "ast"
local Parser = terralib.require('terra/tests/lib/parsing')

-- Global precedence table
local precedence = {

	["or"]  = 0,
	["and"] = 1,

	["<"]   = 2,
	[">"]   = 2,
	["<="]  = 2,
	[">="]  = 2,
	["=="]  = 2,
	["~="]  = 2,

	["+"]   = 3,
	["-"]   = 3,

	["*"]   = 4,
	["/"]   = 4,
	["%"]   = 4,

	unary   = 5,
--	uminus  = 5,
--	["not"] = 5,
--	["#"]   = 5,

	["^"]   = 6,
}

local block_terminators = {
    ['end']    = 1,
 	['else']   = 1,
 	['elseif'] = 1,
 	['until']  = 1,
 	['break']  = 1,
}

--[[ Generic expression operator parsing functions: ]]--
local function leftbinary(P, lhs)
	local op  = P:next().type
	local rhs = P:exp(op)
	return BinaryOp:New(lhs, op, rhs)
end

local function rightbinary(P, lhs)
	local op  = P:next().type
	local rhs = P:exp(op, "right")
	return BinaryOp:New(lhs, op, rhs)
end

local function unary (P)
	local op = P:next().type
	local exp = P:exp(precedence.unary)
	return UnaryOp:New(op, exp)
end	

----------------------------------
--[[ Build Liszt Pratt parser ]]--
----------------------------------
lang = { }

--[[ Expression parsing ]]--
lang.exp = Parser.Pratt() -- returns a pratt parser
		   :prefix("-",   unary)
		   :prefix("not", unary)

		   :infix("or",  precedence["or"],  leftbinary)
		   :infix("and", precedence["and"], leftbinary)

		   :infix("<",   precedence["<"],   leftbinary)
		   :infix(">",   precedence[">"],   leftbinary)
		   :infix("<=",  precedence["<="],  leftbinary)
		   :infix(">=",  precedence[">="],  leftbinary)
		   :infix("==",  precedence["=="],  leftbinary)
		   :infix("~=",  precedence["~="],  leftbinary)

		   :infix("*",   precedence['*'],   leftbinary)
		   :infix('/',   precedence['/'],   leftbinary)
		   :infix('%',   precedence['%'],   leftbinary)

		   :infix("+",   precedence["+"],   leftbinary)
		   :infix("-",   precedence["-"],   leftbinary)
		   :infix('^',   precedence['^'],   rightbinary)
		   :prefix(Parser.default, function(P) return P:simpleexp() end)

-- tuples are used to represent a sequence of comma-seperated expressions that can
-- be found in function calls and will perhaps be used in for statements
lang.tuple = function (P, exprs)
	exprs = exprs or { }
	exprs[#exprs + 1] = P:exp()
	if P:nextif(",") then
		return P:tuple(exprs)
	else 
		return Tuple:New(unpack(exprs))
	end
end

--[[ recursively checks to see if lhs is part of a field index 
     or a table lookup. lhs parameter is already an LValue 
--]]
lang.lvaluehelper = function (P, lhs)
	local cur = P:cur()
	-- table indexing:
	if cur.type == '.' or cur.type == ':' then
		local op = P:next().type
		-- check to make sure the table is being indexed by a valid name
		if not P:matches(P.name) then P:error("expected name after '.'") end
		return P:lvaluehelper(TableLookup:New(lhs, op, Name:New(P:next().value)))

	-- field index / function call?
	elseif P:nextif('(') then
		local args = P:tuple()
		P:expect(')')
		return P:lvaluehelper(Call:New(lhs, args))
	end
	return lhs
end

--[[ This function finds the first name of the lValue, then calls
     lValueHelper to recursively check to table or field lookups. 
--]]
lang.lvalue = function (P)
	if not P:matches(P.name) then
		local token = P:next()
		P:error("Expected name at " .. token.linenumber .. ":" .. token.offset)
	end

	return P:lvaluehelper(Name:New(P:next().value))
end

--[[  simpleexp is called when exp cannot parse the next token...
      we could be looking at an LValue, Value, or parenthetically
      enclosed expression  
--]]
lang.simpleexp = function(P)
	-- catch values
	if P:matches(P.number) then
		return Number:New(P:next().value)		

	-- catch bools
	elseif P:matches('true') or P:matches('false') then
		return Bool:New(P:next().type)

	-- catch parenthesized expressions
	elseif P:nextif("(") then
		local v = P:exp()
		P:expect(")")
		return v
	end

	-- expect LValue
	return P:lvalue()
end

lang.liszt_kernel = function (P)
	-- parse liszt_kernel keyword and argument:
	P:expect("liszt_kernel")

	-- parse parameter
	P:expect("(")
	local param = P:expect(P.name).value
	P:expect(")")

	-- parse block
	local block = P:block()
	P:expect("end")

	return LisztKernel:New(param, block)
end

--[[ Statement Parsing ]]--
--[[ Supported statements:
      - if statement
      - while statement
      - repeat statement
      - do statement
	  - variable declaration
      - assignment
      - variable initialization
      - expression statement
	  - TODO: for statement
--]]
lang.statement = function (P)

	-- check for initialization/declaration
	if (P:nextif("var")) then
		local name = P:lvalue()
		-- differentiate between initialization and declaration
		if (P:nextif("=")) then
			local expr = P:exp()
			return InitStatement:New(name, expr)
		else
			return DeclStatement:New(name)
		end

	--[[ if statement ]]--
	elseif P:nextif("if") then
		local blocks = { }

		local cond = P:exp()

		P:expect("then")
		local body = P:block()
		blocks[#blocks+1] = CondBlock:New(cond, body)

		-- parse all elseif clauses
		while (P:nextif("elseif")) do
			local cond = P:exp()
			P:expect("then")
			local body = P:block()
			blocks[#blocks+1] = CondBlock:New(cond, body)							
			end

		if (P:nextif('else')) then
			blocks[#blocks+1]=P:block()
		end

		P:expect("end")
		return IfStatement:New(unpack(blocks))

	--[[ while statement ]]--
	elseif P:nextif("while") then
		local condition = P:exp()
		P:expect("do")
		local body = P:block()
		P:expect("end")
		return WhileStatement:New(condition, body)

	-- do block end
	elseif P:nextif("do") then
		local body = P:block()
		P:expect("end")
		return DoStatement:New(body)

	-- repeat block until condition
	elseif P:nextif("repeat") then
		local body = P:block()
		P:expect("until")
		local condition = P:exp()
		return RepeatStatement:New(condition, body)

	-- TODO: implement for statement
	-- Just a skeleton. NumericFor loops should be of just one type.
	-- GenericFor loops may be of different types.
	-- What for loops to support within the DSL?
    elseif P:nextif("for") then
		local iterator = P:expect(P.name).value
		if (P:nextif("in")) then
                        local set = P:lvalue()
                        P:expect("do")
                        local body = P:block()
                        P:expect("end")
                        -- ?? what kinds should these be
			return GenericFor:New(iterator, set, body)
		else
			P:expect("=")
                        local exprs = { }
                        exprs[1] = P:exp()
                        P:expect(',')
                        exprs[2] = P:exp()
                        if P:nextif(',') then
                            exprs[3] = P:exp()
                        end
                        P:expect("do")
                        local body = P:block()
                        P:expect("end")
			return NumericFor:New(iterator, unpack(exprs), body)
		end

    elseif P:nextif("foreach") then
		local iterator = P:expect(P.name).value
                P:expect("in")
                local set = P:lvalue()
                P:expect("do")
                local body = P:block()
                P:expect("end")
		return GenericFor:New(iterator, set, body)

	--[[ expression statement / assignment statement ]]--
	else
		local expr = P:exp()
		if (P:nextif('=')) then
			-- check to make sure lhs is an LValue
			if not expr.isLValue() then P:error("expected LValue before '='") end

			local rhs = P:exp()
			return Assignment:New(expr, rhs)
		else
			return expr
		end
	end
end

lang.block = function (P)
	local statements = { }
	local first = P:cur().type
	while not block_terminators[first] do
		statements[#statements+1] = lang.statement(P)
		first = P:cur().type
	end

	if P:nextif('break') then
		statements[#statements+1] = Break:New()
		-- check to make sure break is the last statement of the block
		local key = P:cur().type
		if not block_terminators[key] then
			P:error("block should terminate after the break statement")
		end

		-- make sure we didn't just hit another break statement
		if key == 'break' then P:error("two consecutive break statements") end

	end

	return Block:New(unpack(statements))
end
