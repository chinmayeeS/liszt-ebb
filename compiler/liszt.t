--this file is used to extend the terra parser with
--the liszt language
--use it like: import "compiler.liszt"

local P = require "compiler.parser"

local lisztlib = terralib.require "compiler.lisztlib"
local specialization = terralib.require "compiler.specialization"
local semant = terralib.require "compiler.semant"
-- include liszt library for programmer
L = lisztlib

local lisztlanguage = {
	name        = "liszt", -- name for debugging
	entrypoints = {"liszt"},
	keywords    = {"var", "kernel", "quote", "max", "min"},

	expression = function(self, lexer)
		local ast = P.Parse(lexer)
        if ast.kind == 'LisztKernel' then
            return function (env_fn) 
                local env = env_fn()
                return lisztlib.NewKernel(ast, env)
            end
        elseif ast.kind == 'UserFunction' then
            return function (env_fn)
                local env = env_fn()
                return lisztlib.NewUserFunc(ast, env)
            end
        else -- quote
            return function (env_fn)
                local env = env_fn()

                local specialized   = specialization.specialize(env, ast)
                local checked       = semant.check(env, specialized)

                return checked
            end
        end
	end
}

return lisztlanguage
