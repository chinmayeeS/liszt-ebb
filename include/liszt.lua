-- import runtime, privately to this module (so that it is not exposed to liszt application programmers)
terralib.require('runtime/liszt')
local runtime = runtime
_G.runtime    = nil

local LisztObj = { }

--[[
-- data_type represents type of elements for vectors/ sets/ fields.
-- topo_type represents the type of the topolgical element for a field over
-- some topological set.
-- data_type is int/ float for vectors, vertex/ edge/ face/ cell for
-- topological sets, and ObjType for fields. topo_type is always vertex/ edge/
-- face/ cell. ObjType contains strings for types ("int" and "float") and not
-- the actual type (int or float), since the type could be a vector too.
--]]

--[[ String literals ]]--
local NOTYPE = 'notype'
local TABLE  = 'table'
local INT    = 'int'
local FLOAT  = 'float'
local VECTOR = 'vector'
local VERTEX = 'vertex'
local EDGE   = 'edge'
local FACE   = 'face'
local CELL   = 'cell'

-- need to use this for fields to store the type of field, due to nested types
local ObjType = 
{
    -- type of the object
	obj_type = NOTYPE,
    -- if object consists of elements, then type of elements
	elem_type = NOTYPE,
	-- size of the object (example, vector length)
	size = 0,
}

function ObjType:new()
	local newtype = {}
	setmetatable(newtype, {__index = self})
	newtype.objtype = self.objtype
	newtype.elemtype = self.elemtype
	newtype.size = self.size
	return newtype
end

--[[ Liszt Types ]]--
local TopoElem = setmetatable({kind = "topoelem"}, { __index = LisztObj, __metatable = "TopoElem" })
local TopoSet  = setmetatable({kind = "toposet", data_type = NOTYPE},  { __index = LisztObj, __metatable = "TopoSet" })
local Field    = setmetatable({kind = "field",  topo_type = NOTYPE, data_type = ObjType}, { __index = LisztObj, __metatable = "Field" })
local Scalar   = setmetatable({kind = "scalar", data_type = NOTYPE},                     { __index = LisztObj, __metatable = "Scalar"})
local TopoSet  = setmetatable({kind = "toposet"},  { __index = LisztObj, __metatable = "TopoSet" })

Field  = setmetatable({kind = "field",  topo_type = NOTYPE, data_type = NOTYPE}, { __index = LisztObj, __metatable = "Field" })
Scalar = setmetatable({kind = "scalar", data_type = NOTYPE},                     { __index = LisztObj, __metatable = "Scalar"})

Mesh   = setmetatable({kind = "mesh"},   { __index = LisztObj, __metatable = "Mesh"})
Cell   = setmetatable({kind = "cell"},   { __index = TopoElem, __metatable = "Cell"})
Face   = setmetatable({kind = "face"},   { __index = TopoElem, __metatable = "Face"})
Edge   = setmetatable({kind = "edge"},   { __index = TopoElem, __metatable = "Edge"})
Vertex = setmetatable({kind = "vertex"}, { __index = TopoElem, __metatable = "Vertex"})

DataType = setmetatable({kind = "datatype"}, { __index=LisztObj, __metatable="DataType"})
local DataType   = setmetatable({kind = "datatype"}, { __index=LisztObj, __metatable="DataType"})
Vector           = setmetatable({kind = "vector", data_type = NOTYPE, size = 0},   { __index=DataType})
local VectorType = setmetatable({kind = "vector", data_type = NOTYPE, size = 0}, { __index=DataType})
Vector.__index     = Vector
VectorType.__index = VectorType


local function contains_entry (tbl, entry)
   local set
   pcall(function () set = tbl[entry] ~= nil end)
   return set
end

-------------------------------------------------
--[[ Field methods                           ]]--
-------------------------------------------------
function Field:set_topo_type(topo_type)
   if (type(topo_type) ~= TABLE) then
	   error("Field over unrecognized topological type!!", 3)
   end
   if topo_type == Vertex then
	   self.topo_type = VERTEX
   elseif topo_type == Edge then
	   self.topo_type = EDGE
   elseif topo_type == Face then
	   self.topo_type = FACE
   elseif topo_type == Cell then
	   self.topo_type = CELL
   else
	   error("Field over unrecognized topological type!!", 3)
   end
end

function Field:set_data_type(data_type)
	if data_type == int then
		self.data_type.obj_type = INT
		self.data_type.elem_type = INT
		self.data_type.size = 1
	elseif data_type == float then
		self.data_type.obj_type = FLOAT
		self.data_type.elem_type = FLOAT
		self.data_type.szie = 1
	elseif getmetatable(data_type) == Vector or
		getmetatable(data_type) == VectorType then
		self.data_type.obj_type = VECTOR
		if data_type.data_type == int then
			self.data_type.elem_type = INT
		elseif data_type.data_type == float then
			self.data_type.elem_type = FLOAT
		else
			error("Field over unsupported data type!!", 3)
		end
	   self.data_type.size = data_type.size
   else
	   error("Field over unsupported data type!!", 3)
   end
end

function Field:lField ()
   return self.lfield
end

function Field:lkField ()
   return self.lkfield
end

-------------------------------------------------
--[[ Scalar methods                          ]]--
-------------------------------------------------
function Scalar:lScalar ()
   return self.lscalar
end

function Scalar:lkScalar()
   return self.lkscalar
end


-------------------------------------------------
--[[ Runtime type conversion                 ]]--
-------------------------------------------------
local lElementTypeMap = {
   [Vertex] = runtime.L_VERTEX,
   [Cell]   = runtime.L_CELL,
   [Face]   = runtime.L_FACE,
   [Edge]   = runtime.L_EDGE
}

local lKeyTypeMap = {
   [int]   = runtime.L_INT,
   [float] = runtime.L_FLOAT
}

local function runtimeDataType (data_type)
   if Vector.isVectorType(data_type) then
      return lKeyTypeMap[data_type.data_type], data_type.size
   else
      return lKeyTypeMap[data_type], 1.0
   end
end


-------------------------------------------------
--[[ Vector methods                          ]]--
-------------------------------------------------
function Vector.type (data_type, size)
   if not (contains_entry(lKeyTypeMap, data_type)) then
      error("First argument to Vector.type() should be a Liszt-supported terra data type!", 2)
   end
   if not type(size) == "number" or size < 1 or size % 1 ~= 0 then
      error("Second argument to Vector.type() should be a non-negative integer!", 2)
   end

   return setmetatable({size = size, data_type = data_type}, VectorType)
end

function Vector.new(data_type, size, ...) 
   if not (contains_entry(lKeyTypeMap, data_type)) then
      error("First argument to Vector.new() should be a Liszt-supported terra data type!", 2)
   end
   if not type(size) == "number" or size < 1 or size % 1 ~= 0 then
      error("Second argument to Vector.type() should be a non-negative integer!", 2)
   end

   local init = ...
   local v    = vector(data_type, size)

   -- if the user gave us a table as the third argument, assume it is an array of numeric elements
   if type(init) == 'table' then
      -- copy over first size numeric elements from init
      local max = size > #init and #init or size
      for i = 1, max do
         print(type(init[i]))
         if type(init[i]) ~= 'number' then error("Cannot initialize vector with non-numeric type!", 2) end
         v[i-1] = init[i]
      end

      -- make sure all entries are initialized
      for i = #init, size do
         v[i-1] = 0.0
      end
   else

   end
   return setmetatable({size = size, data_type = data_type, __data = v}, Vector)
end

function Vector.isVector (obj)
   return getmetatable(obj) == Vector
end

function Vector.isVectorType (obj)
   return getmetatable(obj) == VectorType
end

function Vector.add (v1, v2)
   if v1.data_type ~= v2.data_type then
      error("Cannot add vectors of differing types!", 2)
   end
   if v1.size ~= v2.size then
      error("Cannot add vectors of differing lengths!", 2)
   end
end



-------------------------------------------------
--[[ TopoSet methods                         ]]--
-------------------------------------------------
local function toposet_stub (topoelem)
   local tmp = {data_type = topoelem}
   tmp.size = function () return 10 end
   setmetatable(tmp, {__index = TopoSet})
   return tmp
end

Mesh.cells    = toposet_stub(CELL)
Mesh.faces    = toposet_stub(FACE)
Mesh.vertices = toposet_stub(VERTEX)
Mesh.edges    = toposet_stub(EDGE)


-------------------------------------------------
--[[ Mesh methods                            ]]--
-------------------------------------------------
function Mesh:field (topo_type, data_type, initial_val)
   local field = setmetatable({ mesh = self, topo_type = NOTYPE, data_type = ObjType:new() }, { __index = Field })
   field:set_topo_type(topo_type)
   field:set_data_type(data_type)
   local val_type, val_len = runtimeDataType(data_type)
   field.lfield  = runtime.initField(self.ctx, lElementTypeMap[topo_type], val_type, val_len)
   field.lkfield = runtime.getlkField(field.lfield)
   return field
end

function Mesh:fieldWithLabel (topo_type, data_type, label)
   local field = {topo_type = NOTYPE, data_type = ObjType:new(), mesh = self}
   setmetatable(field, { __index = Field })
   field:set_topo_type(topo_type)
   field:set_data_type(data_type)
   local val_type, val_len = runtimeDataType(data_type)
   field.lfield  = runtime.loadField(self.ctx, label, lElementTypeMap[topo_type], val_type, val_len)
   field.lkfield = runtime.getlkField(field.lfield)
   return field
end

function Mesh:scalar (data_type)
   local s    = setmetatable({}, {__index = Scalar })
   s.lscalar  = runtime.initScalar(self.ctx,0,0)
   s.lkscalar = runtime.getlkScalar(s.lscalar)
   return s
end

function Mesh.new () 
   return setmetatable({ }, { __index = Mesh } )
end

LoadMesh = function (filename)
   local mesh = Mesh.new()
   mesh.ctx   = runtime.loadMesh(filename)
   return mesh
end
