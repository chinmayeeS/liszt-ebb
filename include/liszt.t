local T = terralib.require('compiler/types')
local t, Type = T.t, T.Type

local L = {}
local LDB = terralib.require('include/ldb')
L.LDB = LDB

local DECL = terralib.require('include/decl')


local C = DECL.C

-- terra type of a field that refers to another relation
local REF_TYPE    = t.uint
-- terra type of an orientation field
local ORIENT_TYPE = t.uint8


-------------------------------------------------------------------------------
--[[ Export Liszt types:                                                   ]]--
-------------------------------------------------------------------------------
L.int    = t.int
L.uint   = t.uint
L.float  = t.float
L.double = t.double
L.bool   = t.bool
L.vector = t.vector


-------------------------------------------------------------------------------
--[[ Liszt object prototypes:                                              ]]--
-------------------------------------------------------------------------------

-- export object testing routines
L.is_relation = DECL.is_relation
L.is_field    = DECL.is_field
L.is_scalar   = DECL.is_scalar
L.is_vector   = DECL.is_vector
L.is_macro    = DECL.is_macro

local is_vector = L.is_vector

-- local aliases for objects
local LScalar = DECL.LScalar
local LVector = DECL.LVector
local LMacro  = DECL.LMacro

--[[
- An LRelation contains its size and fields as members.  The _index member
- refers to an array of the compressed row values for the index field.

- An LField stores its fieldname, type, an array of data, and a pointer
- to another LRelation if the field itself represents relational data.
--]]


-------------------------------------------------------------------------------
--[[ LScalars:                                                             ]]--
-------------------------------------------------------------------------------
function L.NewScalar (typ, init)
    if not T.Type.isLisztType(typ) or not typ:isValueType() then error("First argument to L.NewScalar must be a Liszt expression type", 2) end
    if not T.luaValConformsToType(init, typ) then error("Second argument to L.NewScalar must be an instance of type " .. typ:toString(), 2) end

    local s  = setmetatable({type=typ}, LScalar)
    local tt = typ:terraType()
    s.data   = terralib.cast(&tt, C.malloc(terralib.sizeof(tt)))
    s:setTo(init)
    return s
end


function LScalar:setTo(val)
   if not T.luaValConformsToType(val, self.type) then error("value does not conform to scalar type " .. self.type:toString(), 2) end
      if self.type:isVector() then
          local v     = is_vector(val) and val or L.NewVector(self.type:baseType(), val)
          local sdata = terralib.cast(&self.type:terraBaseType(), self.data)
          for i = 0, v.N-1 do
              sdata[i] = v.data[i+1]
          end
    -- primitive is easy - just copy it over
    else
        self.data[0] = self.type == t.int and val - val % 1 or val
    end
end

function LScalar:value()
    if self.type:isPrimitive() then return self.data[0] end

    local ndata = {}
    local sdata = terralib.cast(&self.type:terraBaseType(), self.data)
    for i = 1, self.type.N do ndata[i] = sdata[i-1] end

    return L.NewVector(self.type:baseType(), ndata)
end


-------------------------------------------------------------------------------
--[[ LVectors:                                                             ]]--
-------------------------------------------------------------------------------
function L.NewVector(dt, init)
    if not (T.Type.isLisztType(dt) and dt:isPrimitive()) then
        error("First argument to L.NewVector() should "..
              "be a primitive Liszt type", 2)
    end
    if not is_vector(init) and #init == 0 then
        error("Second argument to L.NewVector should either be "..
              "an LVector or an array", 2)
    end
    local N = is_vector(init) and init.N or #init
    if not T.luaValConformsToType(init, L.vector(dt, N)) then
        error("Second argument to L.NewVector() does not "..
              "conform to specified type", 2)
    end

    local data = {}
    if is_vector(init) then init = init.data end
    for i = 1, N do
        -- convert to integer if necessary
        data[i] = dt == L.int and init[i] - init[i] % 1 or init[i] 
    end

    return setmetatable({N=N, type=t.vector(dt,N), data=data}, LVector)
end

function LVector:__codegen ()
    local v1, v2, v3 = self.data[1], self.data[2], self.data[3]
    local v4, v5, v6 = self.data[4], self.data[5], self.data[6]
    local btype = self.type:terraBaseType()

    if     self.N == 2 then
        return `vectorof(btype, v1, v2)
    elseif self.N == 3 then
        return `vectorof(btype, v1, v2, v3)
    elseif self.N == 4 then
        return `vectorof(btype, v1, v2, v3, v4)
    elseif self.N == 5 then
        return `vectorof(btype, v1, v2, v3, v4, v5)
    elseif self.N == 6 then
        return `vectorof(btype, v1, v2, v3, v4, v5, v6)
    end

    local s = symbol(self.type:terraType())
    local t = symbol()
    local q = quote
        var [s]
        var [t] = [&btype](&s)
    end

    for i = 1, self.N do
        local val = self.data[i]
        q = quote 
            [q] 
            @[t] = [val]
            t = t + 1
        end
    end
    return quote [q] in [s] end
end

function LVector.__add (v1, v2)
    if not is_vector(v1) or not is_vector(v2) then
        error("Cannot add non-vector type to vector", 2)
    elseif v1.N ~= v2.N then
        error("Cannot add vectors of differing lengths", 2)
    elseif v1.type == t.bool or v2.type == t.bool then
        error("Cannot add boolean vectors", 2)
    end

    local data = { }
    local tp = T.type_meet(v1.type:baseType(), v2.type:baseType())

    for i = 1, #v1.data do
        data[i] = v1.data[i] + v2.data[i]
    end
    return L.NewVector(tp, data)
end

function LVector.__sub (v1, v2)
    if not is_vector(v1) then
        error("Cannot subtract vector from non-vector type", 2)
    elseif not is_vector(v2) then
        error("Cannot subtract non-vector type from vector", 2)
    elseif v1.N ~= v2.N then
        error("Cannot subtract vectors of differing lengths", 2)
    elseif v1.type == bool or v2.type == bool then
        error("Cannot subtract boolean vectors", 2)
    end

    local data = { }
    local tp = T.type_meet(v1.type:baseType(), v2.type:baseType())

    for i = 1, #v1.data do
        data[i] = v1.data[i] - v2.data[i]
    end

    return L.NewVector(tp, data)
end

function LVector.__mul (a1, a2)
    if is_vector(a1) and is_vector(a2) then
        error("Cannot multiply two vectors", 2)
    end
    local v, a
    if is_vector(a1) then   v, a = a1, a2
    else                    v, a = a2, a1 end

    if     v.type:isLogical()  then
        error("Cannot multiply a non-numeric vector", 2)
    elseif type(a) ~= 'number' then
        error("Cannot multiply a vector by a non-numeric type", 2)
    end

    local tm = t.float
    if v.type == int and a % 1 == 0 then tm = t.int end

    local data = {}
    for i = 1, #v.data do
        data[i] = v.data[i] * a
    end
    return L.NewVector(tm, data)
end

function LVector.__div (v, a)
    if     is_vector(a)    then error("Cannot divide by a vector", 2)
    elseif v.type:isLogical()  then error("Cannot divide a non-numeric vector", 2)
    elseif type(a) ~= 'number' then error("Cannot divide a vector by a non-numeric type", 2)
    end

    local data = {}
    for i = 1, #v.data do
        data[i] = v.data[i] / a
    end
    return L.NewVector(t.float, data)
end

function LVector.__mod (v1, a2)
    if is_vector(a2) then error("Cannot modulus by a vector", 2) end
    local data = {}
    for i = 1, v1.N do
        data[i] = v1.data[i] % a2
    end
    local tp = T.type_meet(v1.type:baseType(), t.float)
    return L.NewVector(tp, data)
end

function LVector.__unm (v1)
    if v1.type:isLogical() then error("Cannot negate a non-numeric vector", 2) end
    local data = {}
    for i = 1, #v1.data do
        data[i] = -v1.data[i]
    end
    return L.NewVector(v1.type, data)
end

function LVector.__eq (v1, v2)
    if v1.N ~= v2.N then return false end
    for i = 1, v1.N do
        if v1.data[i] ~= v2.data[i] then return false end
    end
    return true
end


-------------------------------------------------------------------------------
--[[ LMacros:                                                              ]]--
-------------------------------------------------------------------------------
function L.NewMacro(generator)
    return setmetatable({genfunc=generator}, LMacro)    
end



return L
