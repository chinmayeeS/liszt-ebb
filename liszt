#!terra/bin/terra -mg

-- unfortunately, this script will only work when invoked from the
-- directory in which it's located.  =(  The tricky part is that
-- Terra would have to be installed in a more globally visible way
-- to avoid this...

-- If we can run from elsewhere, then make sure to modify the Lua/Terra paths


local USE_LEGION = false
local USE_GPU    = false
local RUN_SCRIPT = false
local NO_DEBUG   = false
local USE_SPY    = false
local USE_PROF   = false

-- Pre-Process arguments
if #arg > 0 then
  local write_i = 1
  for i=1,#arg do
    if arg[i] == '-legion' or arg[i] == '--legion' then
      USE_LEGION = true
    elseif arg[i] == '-gpu' or arg[i] == '--gpu' then
      USE_GPU = true
    elseif arg[i] == '-nodebug' or arg[i] == '--nodebug' then
      NO_DEBUG = true
    elseif arg[i] == '-spy' or arg[i] == '--spy' then
      USE_SPY = true
    elseif arg[i] == '-prof' or arg[i] == '--prof' then
      USE_PROF = true
    else
      arg[write_i] = arg[i]
      write_i = write_i + 1
    end
  end
  for k=write_i,#arg do arg[k] = nil end
end
if #arg > 0 then RUN_SCRIPT = true end


if USE_SPY then
  if NO_DEBUG then
    error('Must run without --nodebug to use legion spy')
  end
  rawset(_G, 'LISZT_LEGION_USE_SPY', true)
end
if USE_PROF then
  if not NO_DEBUG then
    error('Must run with option --nodebug to use legion prof')
  end
  rawset(_G, 'LISZT_LEGION_USE_PROF', true)
end


if USE_LEGION then
  -- Here, make sure that the legion library is loaded.
  local legion_dir    = './legion'
  local bindings_dir  = legion_dir..'/bindings/terra'
  local runtime_dir   = legion_dir..'/runtime'

  -- Link in a particular library
  if NO_DEBUG then
    terralib.linklibrary(bindings_dir..'/liblegion_terra_release.so')
  else
    terralib.linklibrary(bindings_dir..'/liblegion_terra_debug.so')
  end
  -- Extend the Terra path that looks for C header files
  terralib.includepath = terralib.includepath..';'..runtime_dir..
                                               ';'..bindings_dir
  -- Extend the Lua path
  package.path = package.path .. ';'..bindings_dir..'/?.t'
end

if USE_GPU then
  rawset(_G, 'LISZT_USE_GPU_SIGNAL', true)
end


if RUN_SCRIPT then
  if USE_LEGION then
    assert(terralib.loadfile('./compiler/launch_legion.t'))()
  else
    assert(terralib.loadfile('./compiler/launch_script.t'))()
  end
else
  os.execute('terra/terra')
end
