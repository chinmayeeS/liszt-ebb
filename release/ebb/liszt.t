--this file is used to extend the terra parser with
--the liszt language
--use it like: import "ebb.src.liszt"

local P = require "ebb.src.parser"

local lisztlib = require "ebb.src.lisztlib"
local specialization = require "ebb.src.specialization"
local semant = require "ebb.src.semant"
-- include liszt library for programmer
L = lisztlib

local statement = function(self, lexer)
    local ast, assign = P.ParseStatement(lexer)
    local constructor
    if ast.kind == 'UserFunction' then
        constructor = function (env_fn)
            local env = env_fn()
            return lisztlib.NewUserFunc(ast, env)
        end
    else -- quote
        error("Expected liszt function!", 2)
    end
    return constructor, assign
end

local lisztlanguage = {
    name        = "liszt", -- name for debugging
    entrypoints = {"liszt"},
    -- including max and min as keywords is necessary to get
    -- the parser to interpret them as operators.  This has the
    -- unfortunate affect of not allowing anyone to use 'min' or 'max
    -- as variable names within Liszt code.
    keywords    = {
        "quote",
        "max", "min",
        "var",
        "insert", "into", "delete",
        "_", -- want to hold onto this token in case it's useful
    },

    -- Liszt quotes and anonymous functions
    expression = function (self, lexer)
        local ast = P.ParseExpression(lexer)
        if ast.kind == 'UserFunction' then
            return function (env_fn)
                local env = env_fn()
                return lisztlib.NewUserFunc(ast, env)
            end
        else -- quote
            return function (env_fn)
                local env = env_fn()
                local specialized = specialization.specialize(env, ast)
                local checked     = semant.check(specialized)
                return checked
            end
        end
    end,

    -- named Liszt functions
    statement = statement,
    localstatement = statement,
}

return lisztlanguage
