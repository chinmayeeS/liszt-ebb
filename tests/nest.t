import "compiler.liszt"

local ioOff = L.require 'domains.ioOff'
local mesh  = ioOff.LoadTrimesh('tests/octa.off')

mesh.vertices:NewField('field', L.float)
mesh.vertices.field:Load(1)
local count = L.Global(L.float, 0)

local test_for = liszt (v : mesh.vertices)
	for e in v.edges do
	  count += 1
	end
end

mesh.vertices:map(test_for)

assert(count:get() == 24)
