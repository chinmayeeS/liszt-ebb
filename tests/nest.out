0 0 0
0 0 2
0 0 3
0 0 1
1 1 4
1 1 6
1 1 7
1 1 5
2 2 0
2 2 4
2 2 5
2 2 1
3 3 2
3 3 6
3 3 7
3 3 3
4 4 0
4 4 4
4 4 6
4 4 2
5 5 1
5 5 5
5 5 7
5 5 3
